### ISSUE TITLE
in some cases, GAP must be built with `-pthread` in LDFLAGS

### ISSUE THREAD

**Issue by dimpase** [gap-system/gap#5192]
### Observed behaviour

On Fedora 34, 
`Semigroups` package test `SemigroupsTestAll();`
fails with `Error, Enable multithreading to use std::thread: Operation not permitted`

### Expected behaviour

it should pass

-----

This is GAP 4.12.1 built with 
```
$ gcc -v
Using built-in specs.
COLLECT_GCC=/usr/bin/gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/11/lto-wrapper
OFFLOAD_TARGET_NAMES=nvptx-none
OFFLOAD_TARGET_DEFAULT=1
Target: x86_64-redhat-linux
Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,objc,obj-c++,ada,go,d,lto --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --with-linker-hash-style=gnu --enable-plugin --enable-initfini-array --with-isl=/builddir/build/BUILD/gcc-11.3.1-20220421/obj-x86_64-redhat-linux/isl-install --enable-offload-targets=nvptx-none --without-cuda-driver --enable-gnu-indirect-function --enable-cet --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 11.3.1 20220421 (Red Hat 11.3.1-2) (GCC) 
```
on x86_64 running Fedora 34, with ld `GNU ld version 2.35.2-6.fc34`. 
Configuring GAP with `./configure LDFLAGS=-pthread <...>` cures the issue.

This appears to be platform-specific, e.g. on current Gentoo (with gcc 12.2.0, ld 2.39.0)  the same test passes.

See https://github.com/semigroups/Semigroups/issues/771#issuecomment-1304459598 for more details.

**Comment by dimpase** [comment]
If we decide it's worth a fix, something like https://www.gnu.org/software/autoconf-archive/ax_pthread.html may be used.

**Comment by fingolfin** [comment]
We already are using `AX_PTHREAD`, and we already add the flags it produces to our `LDFLAGS`. To debug what is going on here, we need more information from you:

- first off, to double check: you are compiling GAP yourself from pristine GAP 4.12.1 sources, yes? (i.e. not using a system package)
- please add the content of `sysinfo.gap` (the one in the directory GAP was built, *NOT* the one installed by `make install`)
- please attach `config.log`
- please attach `GNUmakefile`

**Comment by dimpase** [comment]
I wonder whether you by default expect GAP to be linked with libpthread ? Because in two cases I have looked (Fedora 34 and Gentoo) this does not happen. However, the bug shows only on Fedora. So it must be a difference in toolchains.

GAP is built from 4.12.1 sources in all cases.

**Comment by fingolfin** [comment]
We could perhaps figure it out if you provide the data I asked for...

**Comment by dimpase** [comment]
> We already are using `AX_PTHREAD`

you use it conditionally: confugure.ac says:
```
dnl pthreads
AS_IF([test "x$enable_hpcgap" = xyes -o "x$with_julia" != xno ],[
  AX_PTHREAD
  ])
```
As I don't enable julia or hpcgap, I don't see how it ever enters the play.

Here is the `sysconfig.gap`,  on the system I see the bug:
```
# This file has been generated by the GAP build system,
# DO NOT EDIT MANUALLY!

# This file exists for use by the GAP compiler `gac` and also by the build
# systems of GAP packages.
#
# It is deliberately written in a way that makes it usable from shell scripts
# *and* Makefiles: It only consists of comment lines like this one, starting
# with a '#'; blank linkes; or lines containing assignments of the form
#  VARIABLE_NAME="CONTENT"


# GAParch: the "architecture" of this GAP build. First comes the result of
# `config.guess`. Next one of the strings "-default", "-hpcgap", "-julia" to
# indicate various major build variants of GAP. Then the ABI (32 or 64). Then
# a string of the form '-kvNNN' where NNN is the GAP kernel major version.
#
# Used to decide where to load package binaries from (to support installing
# binaries for multiple architectures in parallel on a single computer).
GAParch=x86_64-pc-linux-gnu-default64-kv8

# GAP_ABI: indicates whether GAP was built in 32 or 64 bit mode.
# Deprecated, may be removed in a future GAP release.
# Possible values: 32, 64
GAP_ABI=64

# GAP_HPCGAP: indicates whether HPC-GAP support was compiled in.
# Possible values: yes, no
GAP_HPCGAP=no

# GAP_VERSION: GAP version.
GAP_VERSION="4.12.1"

# GAP_BUILD_VERSION: GAP build version. For releases, this is identical to
# GAP_VERSION. For development builds, it is the version followed by extra
# data indicating which git commit the build was based on
GAP_BUILD_VERSION="4.12.1"

# GAP_KERNEL_{MAJOR,MINOR}_VERSION: the GAP kernel version (both integer values).
GAP_KERNEL_MAJOR_VERSION=8
GAP_KERNEL_MINOR_VERSION=0

# GAP: path to the `gap` executable (may be a shell script)
GAP="/home/scratch/scratch2/dimpase/sage/sagetrac-mirror/upstream/gap-4.12.1/gap"

# GAC: path to the `gac` executable (may be a shell script)
GAC="/home/scratch/scratch2/dimpase/sage/sagetrac-mirror/upstream/gap-4.12.1/gac"

# File extension used by object files created by GAC
GAP_OBJEXT=o

# Build flags for use by `gac` resp. package build systems: these correspond to
# the usual variables as used by e.g. autotools, with the added prefix `GAP_`.
GAP_CC="gcc "
GAP_CXX="g++ -std=gnu++11 "
GAP_CFLAGS="  -g -O2"
GAP_CXXFLAGS="  -g -O2"
GAP_CPPFLAGS="-I/home/scratch/scratch2/dimpase/sage/sagetrac-mirror/upstream/gap-4.12.1/build -I/home/scratch/scratch2/dimpase/sage/sagetrac-mirror/upstream/gap-4.12.1/src -I/home/scratch/scratch2/dimpase/sage/sagetrac-mirror/upstream/gap-4.12.1 -DUSE_GASMAN=1  "
GAP_LDFLAGS=" "
GAP_LIBS="-lm -ldl -lutil "

# Extra flags for use by `gac`
GAC_CFLAGS="-fPIC"
GAC_LDFLAGS="-shared"

# For packages: the search prefix for GMP used by GAP
# (this is left empty when none was specified, or when
# GAP is installed via 'make install')
GMP_PREFIX=""
```

**Comment by dimpase** [comment]
if in configure.ac I invoke `AX_PTHREAD` unconditionally, then `gap` gets linked to libpthread (and the bug is gone), otherwise not.

Perhaps one should just make calling  `AX_PTHREAD` unconditional?

**Comment by fingolfin** [comment]
Indeed, you are correct (and I totally missed this when I looked at the code): while we add pthread flags unconditionally in `Makefile.rules` via
```
GAP_CFLAGS += $(PTHREAD_CFLAGS)
SYSINFO_CFLAGS += $(PTHREAD_CFLAGS)
...
GAP_CXXFLAGS += $(PTHREAD_CFLAGS)
SYSINFO_CXXFLAGS += $(PTHREAD_CFLAGS)
...
GAP_LDFLAGS += $(PTHREAD_CFLAGS) $(PTHREAD_LIBS)

```
this does not suffice in general, because as you point out, `AX_PTHREAD` is only run for HPC-GAP and Julia builds.

The reason why the pthread flags were nevertheless visible on my Ubuntu machine: apparently at some point in the (distant) past (well distant enough that I have no recollection of this), I seem to have run `./configure LDFLAGS="STUFF"` (where `STUFF` included `-pthread`) in my primary clone on my Ubuntu server. And this has stuck around till today (across many re-runs of `autogen.sh` and `configure` -- perhaps I should run `make distclean` from time to time). I may even have done this as part of an investigation into https://github.com/semigroups/Semigroups/issues/771. Whatever the source, I've now removed this.

Alas, even then it *still* works fine for me on Ubuntu; so I can't reproduce the problem at all. I may have to set up a Fedora VM...
```
$ ldd gap
	linux-vdso.so.1 (0x00007ffdc1b45000)
	libgmp.so.10 => /lib/x86_64-linux-gnu/libgmp.so.10 (0x00007f8ccd7d9000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f8ccd7bd000)
	libreadline.so.8 => /lib/x86_64-linux-gnu/libreadline.so.8 (0x00007f8ccd76d000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8ccd61e000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f8ccd618000)
	libutil.so.1 => /lib/x86_64-linux-gnu/libutil.so.1 (0x00007f8ccd613000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f8ccd41f000)
	libtinfo.so.6 => /lib/x86_64-linux-gnu/libtinfo.so.6 (0x00007f8ccd3ef000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f8cce6e0000)
$ ldd pkg/semigroups/bin/x86_64-pc-linux-gnu-default64-kv8/semigroups.so
	linux-vdso.so.1 (0x00007ffdf0fa8000)
	libsemigroups.so.2 => /home/mhorn/Projekte/GAP/gap/pkg/semigroups/libsemigroups/../bin/lib/libsemigroups.so.2 (0x00007f5a4270d000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f5a426d8000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f5a424f6000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f5a42304000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f5a422e9000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f5a4219a000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f5a4375a000)
```
So in the above, no libpthread or anyting like that for GAP, but semigroups uses it (also libsemigroups), and `SemigroupsTestAll();` passes.


Anyway, the way I see it, this is *not* a bug in GAP but rather a deficiency in the design of libgcc (?): forcing a program which is not multithreaded to use a linker option to enable multithreading it never needs, just in case some runtime extension aka "plugin" might want to use it, seems backwards. Alas, if that's the only way to do it, we can activate this workaround (but it should be explicitly documented as such, so that if someone wonders why we link against pthreads unconditionaly this becomes clear.

**Comment by dimpase** [comment]
Indeed, as I already said it's platform-specific, as I can't reproduce it on Gentoo with a newer gcc/ld/etc toolchain.
IIRC, also having Semigroups to use its vendored libsemigroups, rather than an external one, fixes the issue.  

Unfortunately I don't know how to test for this deficiency easily. 

I could also say that having threads involved in a plugin might require the whole thing to be thread-safe,
especially thread-local variables must be appropriately declared. If these declarations are conditional on appropriate
CPPFLAGS, one might get hard to debug problems.

PS. Setting up a Fedora VM in Docker (not that I like Docker, but at least on macOS it works OK) 
takes about a minute -  on a fast internet.

**Comment by dimpase** [comment]
On the affected system also the usual GAP way of building GAP and Semigroups (with its dependencies) demonstrates the bug.

**Comment by dimpase** [comment]
By the way, the following way to start GAP also makes the test pass:
```
$ LD_PRELOAD=/lib64/libpthread.so.0 ./gap
```

**Comment by fingolfin** [comment]
Resolved by #5504
